### PR TITLE
feat: Add PubSub module

### PR DESCRIPTION
**Pull request by dejandjenic**
This PR add support for GCP PubSub service. I added test for testing this functionality

### CODE REVIEW

File: src/Testcontainers.PubSub/PubSubBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+namespace Testcontainers.PubSub;
+
+/// <inheritdoc cref="ContainerBuilder{TBuilderEntity, TContainerEntity, TConfigurationEntity}" />
+[PublicAPI]
+public sealed class PubSubBuilder : ContainerBuilder<PubSubBuilder, PubSubContainer, PubSubConfiguration>
+{
+    public const string PubSubPort = "8085";
+    const string Image = "gcr.io/google.com/cloudsdktool/google-cloud-cli:446.0.1-emulators";
```
Comment (HofmeisterAn) [inline src/Testcontainers.PubSub/PubSubBuilder.cs]:
```suggestion
    public const string GoogleCloudCliImage = "gcr.io/google.com/cloudsdktool/google-cloud-cli:446.0.1-emulators";

    public const ushort PubSubPort = 8085;
```

File: src/Testcontainers.PubSub/PubSubBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    /// </summary>
+    /// <param name="resourceConfiguration">The Docker resource configuration.</param>
+    private PubSubBuilder(PubSubConfiguration resourceConfiguration)
+        : base(resourceConfiguration)
+    {
+        DockerResourceConfiguration = resourceConfiguration;
+    }
+
+    // /// <inheritdoc />
+    // protected override PubSubConfiguration DockerResourceConfiguration { get; }
```
Comment (HofmeisterAn) [inline src/Testcontainers.PubSub/PubSubBuilder.cs]: Could you please tidy up the class (`PubSubConfiguration`, ...) in alignment with the other modules? I think the Firestore module is a good example (removing unnecessary comments, indentation, etc.).

File: src/Testcontainers.PubSub/PubSubBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    protected override PubSubBuilder Init()
+    {
+        return base.Init()
+                .WithImage(Image)
+                .WithPortBinding(PubSubPort,true)
+                .WithWaitStrategy(Wait.ForUnixContainer().UntilMessageIsLogged("started."))
+                .WithEntrypoint("gcloud")
+                .WithCommand("beta", "emulators", "pubsub", "start", "--host-port", "0.0.0.0:" + PubSubPort) 
+            ;
```
Comment (HofmeisterAn) [inline src/Testcontainers.PubSub/PubSubBuilder.cs]:
```suggestion
            .WithImage(GoogleCloudCliImage)
            .WithPortBinding(PubSubPort, true)
            .WithEntrypoint("gcloud")
            .WithCommand("beta", "emulators", "pubsub", "start", "--host-port", "0.0.0.0:" + PubSubPort)
            .WithWaitStrategy(Wait.ForUnixContainer().UntilMessageIsLogged("(?s).*running.*$"));
```

File: tests/Testcontainers.PubSub.Tests/Testcontainers.PubSub.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5" />
+    <PackageReference Include="xunit" Version="2.4.2" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(SolutionDir)src/Testcontainers.PubSub/Testcontainers.PubSub.csproj" />
+    <ProjectReference Include="$(SolutionDir)tests/Testcontainers.Commons/Testcontainers.Commons.csproj" />
+  </ItemGroup>
+
+</Project>
```
Comment (HofmeisterAn) [inline tests/Testcontainers.PubSub.Tests/Testcontainers.PubSub.Tests.csproj]:
```suggestion
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFrameworks>net6.0</TargetFrameworks>
        <IsPackable>false</IsPackable>
        <IsPublishable>false</IsPublishable>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2"/>
        <PackageReference Include="coverlet.collector" Version="6.0.0"/>
        <PackageReference Include="xunit.runner.visualstudio" Version="2.5.0"/>
        <PackageReference Include="xunit" Version="2.5.0"/>
        <PackageReference Include="Google.Cloud.PubSub.V1" Version="3.5.0" />
    </ItemGroup>
    <ItemGroup>
        <ProjectReference Include="$(SolutionDir)src/Testcontainers.PubSub/Testcontainers.PubSub.csproj"/>
        <ProjectReference Include="$(SolutionDir)tests/Testcontainers.Commons/Testcontainers.Commons.csproj"/>
    </ItemGroup>
</Project>
```

File: src/Testcontainers.PubSub/PubSubContainer.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    /// Initializes a new instance of the <see cref="PubSubContainer" /> class.
+    /// </summary>
+    /// <param name="configuration">The container configuration.</param>
+    /// <param name="logger">The logger.</param>
+    public PubSubContainer(PubSubConfiguration configuration, ILogger logger)
+        : base(configuration, logger)
+    {
+    }
+
+    public string GetEmulatorEndpoint() => new UriBuilder(Uri.UriSchemeHttp,Hostname,GetMappedPublicPort(PubSubBuilder.PubSubPort)).ToString();
```
Comment (HofmeisterAn) [inline src/Testcontainers.PubSub/PubSubContainer.cs]:
```suggestion
    /// <summary>
    /// Gets the PubSub emulator endpoint.
    /// </summary>
    /// <returns>The PubSub emulator endpoint.</returns>
    public string GetEmulatorEndpoint()
    {
        return new UriBuilder(Uri.UriSchemeHttp, Hostname, GetMappedPublicPort(PubSubBuilder.PubSubPort)).ToString();
    }
```

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Thank you for your PR. I have a couple of minor requests to ensure consistency with all the other modules.

File: src/Testcontainers.PubSub/PubSubBuilder.cs (hunk 1 above)
Comment (dejandjenic) [reply src/Testcontainers.PubSub/PubSubBuilder.cs]: done

File: src/Testcontainers.PubSub/PubSubBuilder.cs (hunk 2 above)
Comment (dejandjenic) [reply src/Testcontainers.PubSub/PubSubBuilder.cs]: done

File: src/Testcontainers.PubSub/PubSubBuilder.cs (hunk 3 above)
Comment (dejandjenic) [reply src/Testcontainers.PubSub/PubSubBuilder.cs]: done

File: src/Testcontainers.PubSub/PubSubContainer.cs (hunk 5 above)
Comment (dejandjenic) [reply src/Testcontainers.PubSub/PubSubContainer.cs]: done

**Comment by dejandjenic** [comment]
@HofmeisterAn can you pls review this again

File: tests/Testcontainers.PubSub.Tests/Testcontainers.PubSub.Tests.csproj (hunk 4 above)
Comment (dejandjenic) [reply tests/Testcontainers.PubSub.Tests/Testcontainers.PubSub.Tests.csproj]: done

File: tests/Testcontainers.PubSub.Tests/PubSubHelper.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      };
+    }
+
+    public static async Task<PullResponse> PullAsyncImmediatelly(this SubscriberServiceApiClient subscriber,
+      SubscriptionName subscription, int maxMessages = 100)
+    {
+      return await subscriber.PullAsync(new PullRequest
+      {
+        SubscriptionAsSubscriptionName = subscription,
+        ReturnImmediately = true,
```
Comment (HofmeisterAn) [inline tests/Testcontainers.PubSub.Tests/PubSubHelper.cs]: Thanks for addressing the requests, can you please that a look at this obsolete property, it says:

> 'PullRequest.ReturnImmediately' is obsolete

I am happy to finalize and merge the PR afterward.

**Comment by dejandjenic** [comment]
done

File: tests/Testcontainers.PubSub.Tests/PubSubHelper.cs (hunk 6 above)
Comment (dejandjenic) [reply tests/Testcontainers.PubSub.Tests/PubSubHelper.cs]: done

**Review by HofmeisterAn — APPROVED** [review]

**Pull request merged**